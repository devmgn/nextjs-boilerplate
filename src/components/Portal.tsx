'use client';

import { useState } from 'react';
import { createPortal } from 'react-dom';
import { useLifecycles } from 'react-use';

type PortalProps = React.PropsWithChildren<{
  disablePortal?: boolean;
  container?: Parameters<typeof createPortal>[1];
}>;

const Portal: React.FC<PortalProps> = ({
  disablePortal,
  container: _container,
  children,
}) => {
  const [container, setContainer] = useState<PortalProps['container'] | null>(
    null,
  );

  useLifecycles(
    () => setContainer(_container ?? document.body),
    () => setContainer(null),
  );

  if (disablePortal) {
    return children;
  }

  return container ? createPortal(children, container) : null;
};

export default Portal;
