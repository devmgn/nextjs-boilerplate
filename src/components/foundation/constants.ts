export const FONT_FAMILY = {
  SANS_SERIF: 'system-ui, -apple-system, sans-serif',
  SERIF: "serif",
  MONOSPACE: 'monospace, monospace',
};

export const MIN_DEVICE_WIDTH = 320;
