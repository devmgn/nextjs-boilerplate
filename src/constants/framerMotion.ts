import type { Transition } from 'framer-motion';

export const TRANSITION_DURATION = 0.3 as const;
export const EASING = 'easeInOut' as const;
export const TRANSITION: Transition = {
  duration: TRANSITION_DURATION,
  ease: EASING,
} as const;

export const VARIANTS = {
  active: { opacity: 1 },
  inactive: { opacity: 0 },
};
