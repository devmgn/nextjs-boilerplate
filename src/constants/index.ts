import type { Transition } from 'framer-motion';

export const ROOT_CONTAINER_ID = '__next';
export const TRANSITION_DURATION = 0.3;
export const EASING = 'easeInOut';
export const TRANSITION: Transition = { duration: TRANSITION_DURATION, ease: EASING };
