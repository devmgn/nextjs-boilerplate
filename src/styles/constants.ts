import { createTheme } from 'styled-breakpoints';

export const BREAKPOINTS = createTheme({
  xs: '375px',
  s: '576px',
  m: '768px',
  l: '992px',
  xl: '1200px',
  xxl: '1400px',
});

export const FONT_FAMILY = {
  SANS_SERIF: 'system-ui, -apple-system, sans-serif',
  SERIF: 'serif',
  MONOSPACE: 'monospace, monospace',
} as const;
