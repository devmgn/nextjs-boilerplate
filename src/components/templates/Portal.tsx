import { useEffect, useState } from 'react';
import ReactDOM from 'react-dom';
import { ROOT_CONTAINER_ID } from '@/constants/appSettings';

type PortalProps = {
  children: React.ReactNode;
};

const Portal: React.FC<PortalProps> = ({ children }) => {
  const [container, setContainer] = useState<HTMLElement | null>(null);

  useEffect(() => {
    setContainer(document.getElementById(ROOT_CONTAINER_ID));
    return () => setContainer(null);
  }, []);

  return container ? ReactDOM.createPortal(children, container) : null;
};

export default Portal;
