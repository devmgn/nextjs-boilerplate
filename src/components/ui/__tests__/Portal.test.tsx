import { render } from '@testing-library/react';
import { ROOT_CONTAINER_ID } from '@/constants';
import Portal from '../Portal';

describe('Portal', () => {
  describe('スナップショットテスト', () => {
    test('コンポーネントをレンダリングする要素があるとき', () => {
      let container: HTMLElement | null = document.createElement('div');
      container.id = ROOT_CONTAINER_ID;
      document.body.appendChild(container);

      const { asFragment } = render(<Portal>children</Portal>, { container: container! });
      expect(asFragment()).toMatchSnapshot();
      container = null;
    });

    test('コンポーネントをレンダリングする要素がないとき', () => {
      const { asFragment } = render(<Portal>children</Portal>);

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
