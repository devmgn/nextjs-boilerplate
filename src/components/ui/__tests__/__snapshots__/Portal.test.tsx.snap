// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal スナップショットテスト コンポーネントをレンダリングする要素があるとき 1`] = `
<DocumentFragment>
  children
</DocumentFragment>
`;

exports[`Portal スナップショットテスト コンポーネントをレンダリングする要素がないとき 1`] = `<DocumentFragment />`;
